vortex-io/src/read_ranges.rs

```diff
@@ -0,0 +1,150 @@
+        Self {
+            read,
+            dispatcher,
+            max_gap,
+        }
+    }
+}
+
+impl<R: VortexReadAt> VortexReadRanges<R> {
+    pub fn read_byte_ranges(
+        &self,
+        ranges: Vec<Range<usize>>,
+    ) -> impl Future<Output = io::Result<Vec<Bytes>>> + 'static {
+        let dispatcher = self.dispatcher.clone();
+        let reader = self.read.clone();
+        let max_gap = self.max_gap;
+        Box::pin(async move {
+            let merged_ranges = merge_ranges(ranges.clone(), max_gap);
+            let read_ranges = stream::iter(merged_ranges.iter().cloned())
+                .map(|r| {
+                    dispatcher
+                        .dispatch({
+                            let reader = reader.clone();
+                            move || async move {
+                                reader
+                                    .read_byte_range(r.start as u64, (r.end - r.start) as u64)
+                                    .await
+                            }
+                        })
+                        .vortex_expect("dispatch async")
+                        .map(|bytes| {
+                            bytes
+                                .map_err(|e| io::Error::new(ErrorKind::Other, e))
+                                .and_then(|b| b)
+                        })
+                })
+                .buffered(MAX_BUFFERED_READS)
+                .try_collect::<Vec<_>>()
+                .await?;
+
+            let mut result_bytes = Vec::with_capacity(ranges.len());
+            for range in ranges {
+                let read_idx = merged_ranges.partition_point(|mr| mr.start <= range.start) - 1;
+
+                let read_range_start = merged_ranges[read_idx].start;
+                let read_bytes = &read_ranges[read_idx];
+                let start = range.start - read_range_start;
+                let end = range.end - read_range_start;
+                result_bytes.push(read_bytes.slice(start..end.min(read_bytes.len())));
+            }
+
+            Ok(result_bytes)
+        })
+    }
+}
+
+fn merge_ranges(mut ranges: Vec<Range<usize>>, max_gap: usize) -> Vec<Range<usize>> {
+    if ranges.is_empty() {
+        return Vec::new();
+    }
+
+    ranges.sort_unstable_by_key(|r| r.start);
+    let mut merged_ranges = Vec::with_capacity(ranges.len());
+
+    let mut start_idx = 0;
+    let mut end_idx = 0;
+
+    while start_idx < ranges.len() {
+        let mut range_end = ranges[start_idx].end;
+
+        while end_idx < ranges.len()
+            && ranges[end_idx]
+                .start
+                .checked_sub(range_end)
+                .map(|gap| gap <= max_gap)
+                .unwrap_or(true)
+        {
+            range_end = range_end.max(ranges[end_idx].end);
+            end_idx += 1;
+        }
+
+        merged_ranges.push(ranges[start_idx].start..range_end);
+        start_idx = end_idx;
```

Comment:
end_idx is inclusive, so I think this should be `start_idx = end_idx + 1`

Reply:
I think you're right. This is repeating the last chunk read twice

Reply:
No, the previous loop stops only if the gap is bigger so once the gap is bigger than the max gap then end_idx is one further than we'd like to merge so it's the next range to consider